### ISSUE TITLE
Print the event page using the browser

### ISSUE THREAD

**Issue by florv** [indico/indico#2408]
Several users want to print the event page with the exact same layout that is displayed on screen, in particular people are interested to print the "Indico style" and "Administrative style" layout.

To achieve this they are currently using the "Print page" feature of their user agent which renders the pages terribly at best thanks to [CSS2.2](https://www.w3.org/TR/CSS22/page.html) being useless (not specific enough, implementation are allowed to do nearly anything, …) and the fact that almost nobody cares to have a printable web. Page breaks are added in different way by each browsers and are sometimes not added properly which ends up with content being not rendered. Eventually [CSS Fragmentation Module Level 3](https://www.w3.org/TR/css-break-3/) will fix many (if not all) of the current issues but it is not yet supported by any released user agent.

We have thought of the following possibilities to satisfies the user's printing wishes:

1. Adapt the current styles to use only old web features that are printed mostly correctly by the supported browsers
2. Render the pages server side and serve a PDF file
3. Print screenshot of the pages

With 1:

`-`  We won't be able keep the fancy layout used for screen display
`-`  We will most likely end up with two versions of the pages to maintain
`-`  Improving the layout will be difficult without making it non-printable
`+` We use the same technology for printing and screen rendering

With 2:

`-`  We will have two versions of the pages to maintain
`-`  We will most likely end up using a different technology to render the print version
`+` Rendering is independent of the user agent
`+` We can use a technology meant to render pages to be printed

With 3:

`-`  The only way to take screenshot is an HTML on canvas renderer [1]
`-`  The screenshot may not be rendered with the correct DPI for printing
`-`  We need to layout the screenshot on multiple page ourselves
`+` We use the same technology for printing and screen rendering

[1] Such as [html2canvas](http://html2canvas.hertzen.com/) which may not implement all the CSS features we use, put positioning of elements should will be extracted from the screen rendering.

**Comment by ThiefMaster** [comment]
Option #4 would be using [PhantomJS](http://phantomjs.org/screen-capture.html) ;)

**Comment by florv** [comment]
IMHO that counts as 2 with a terrible choice of technology ~~meant~~ for prints. :stuck_out_tongue:

**Comment by DirkHoffmann** [comment]
As mentioned on [project-indico-administrators list](https://groups.cern.ch/group/project-indico-administrators/Lists/Archive%2FPDF generation for Contribution), I noticed that it is sufficient to remove the CSS elements concerning/containing **flex** for printing the timetable (in different formats) correctly with Firefox. The browser layout is not noticeably changed by that. 
Maybe that can be a quick fix until all browsers fully support CSS for printing as well? If not for all pages, it seems most important to me to allow correct PDF export of the timetable (in different formats).

I did not understand what the **flex** display is supposed to to exactly (applied to **body**, IIRC), in particular when the orientation remains standard and not *-reversed. Does anybody have a quick answer/reference for that?

**Comment by ThiefMaster** [comment]
I don't remember the reason but it was needed for *something*...

@florv, do you remember it by any chance? IIRC it was you adding it (or @OmeGak?)

**Comment by DirkHoffmann** [comment]
> I don't remember the reason but it was needed for something...

It can be easily verified by removing it from all views/pages and check, if 
something changes. (I would expect the "event home page" to need this for full 
flexibility, but not many more.)

But at least for the timetables (I checked 2 formats, and they are all 
almost pre-CSS, I believe.) adding the CSS-flex could be reverted and allow 
printing with least effort.

**Comment by pferreir** [comment]
@ThiefMaster I wrote the flex layout of the timetable. And it was needed in order to keep headers properly aligned in all possible situations. Maybe removing them doesn't cause great harm, but I'm pretty sure the alignment won't be the same. It would be nice to double-check, though.

**Comment by pferreir** [comment]
BTW, Firefox decides that only the first page of a conference timetable should be printed. E.g. [this one](https://indico.cern.ch/event/523655/timetable/#20160906.detailed) (switch to printing mode and then try to print).

**Comment by DirkHoffmann** [comment]
I tried [your timetable](https://indico.cern.ch/event/523655/timetable/#all), and it prints quite well. I have removed the CSS properties
`body{
display:flex;
flex-direction: column;}` 
**two times** and obtained (via print-to-pdf in Firefox) [this file for the print view](https://github.com/indico/indico/files/456648/allpages.pdf) and [this file for the standard view](https://github.com/indico/indico/files/456647/allpages0.pdf)
The two (identical) CSS come from a file `timetable_*.css` and `screen_sass_*.css` respectively. So that seems to point to a place where one occurence could already be removed rather than keeping the redundancy.
Regarding @pferreir's comment above regarding alignment: Maybe it was needed to add the flex layout to the table. (Maybe it was not :wink:.) But in any case, my suggestion was to remove the flex only for the print layout, not screen or other devices. And whatever the side effect will be (limited to printing), it cannot be worse than only being able to print the uppermost part of the table on a single page (+ one additional empty page by default). Right? :smiley:

**Comment by pferreir** [comment]
> And whatever the side effect will be (limited to printing), it cannot be worse than only being able to print the uppermost part of the table on a single page (+ one additional empty page by default). Right? 😃

Totally agreed 😃

**Comment by pferreir** [comment]
I think there's a misunderstanding. We're talking of different timetables 😄 

You're talking about [conference timetables](https://indico.cern.ch/event/523655/timetable/) and I mean [meeting timetables](https://indico.cern.ch/event/563885/). The former doesn't need flex, the latter depends on it.

I'm not sure how flex is influencing the rendering of conference timetables, it shouldn't be used at all. If it is, then it's probably a bug.

**Comment by DirkHoffmann** [comment]
Today (5:32am) Pedro Ferreira wrote:

> I think there's a misunderstanding. We're talking of different timetables ?
> 
> You're talking about conference timetables and I mean meeting timetables.

Oh yes, we get to an important point there!

> The former doesn't need flex, the latter depends on it.

I did not check for the meeting style. Will do later.

> I'm not sure how flex is influencing the rendering of conference timetables,
> it shouldn't be used at all. If it is, then it's probably a bug.

Good. That means it can be fixed quickly without interfering with other 
features! :-)

**Comment by jouvin** [comment]
One year later the problem is still there, despite being flagged as critical... It would be good to have at least a workaround: I agree with @DirkHoffmann that not being able to print the timetable is the worst thing that can happen! Everything else is an improvement!

**Comment by jouvin** [comment]
BTW, my tests confirm that this is a Firefox-only problem. I tested successfully Safari, Edge and Opera. And Chrome is reported to work by CERN people. Despite what I said in my previous message, I think we need to put some pressure on Firefox to fix this problem (as this is difficult to say that you support everything but Firefox).

**Comment by DirkHoffmann** [comment]
For me it looks like there is a malformed CSS at work in Indico (see https://github.com/indico/indico/issues/2408#issuecomment-248290472 above), as some elements / sequences are duplicated.
Browsers can be more or less tolerant to malformed CSS (or HTML), and it would be new that Firefox is most pettifogging of all browsers. (Opera used to be.) But well, I think both sides need to work. And it is easier to defend that CSS should be well formed than that a browser should be error tolerant.

**Comment by pferreir** [comment]
> One year later the problem is still there, despite being flagged as critical... It would be good to have at least a workaround

I agree, but we cannot solve problems that don't depend solely on us. From all the research we've made, the conclusion was that the only alternative would be rewriting the recently-rewritten timetable code **without** flexbox.

> For me it looks like there is a malformed CSS at work in Indico 

What leads you to conclude this? @florv check the CSS for problems quite extensively and found no indication of malformation on our side. But if you have new clues, please do share them on this thread. We really want to solve this issue and dedicated extensive time to studying options (we even considered using a headless browser on the server side to generate previews).

**Comment by DirkHoffmann** [comment]
> What leads you to conclude this? @florv check the CSS for problems quite extensively and found no indication of malformation on our side. But if you have new clues, please do share them on this thread. We really want to solve this issue and dedicated extensive time to studying options (we even considered using a headless browser on the server side to generate previews).

I don't know, if my proposal above (https://github.com/indico/indico/issues/2408#issuecomment-244934589) is still valid with the present version. Did you check, if those changes would cause new problems in other browsers, in addition to removing those in Firefox? Or why do you think these ideas are old or useless?

**Comment by jouvin** [comment]
Any good reason why this issue is still open after the workaround added by @ThiefMaster in October. According to my tests with 2.0.3, the problem is gone...
